src/middleware/compression.rs

```diff
@@ -0,0 +1,239 @@
+#![cfg(feature = "compression")]
+use brotli2::read::{BrotliDecoder, BrotliEncoder};
+use bytes::{Buf, Bytes, IntoBuf};
+use flate2::read::{DeflateDecoder, DeflateEncoder, GzDecoder, GzEncoder};
+use futures::future::FutureObj;
+use futures::stream::StreamExt;
+use http::{
+    header::{HeaderValue, ACCEPT_ENCODING, CONTENT_ENCODING},
+    status::StatusCode,
+    HeaderMap,
+};
+use http_service::Body;
+use lazy_static::lazy_static;
+use regex::Regex;
+use std::io::Read;
+
+use crate::{
+    middleware::{Middleware, Next},
+    response::IntoResponse,
+    Context, Error, Response,
+};
+
+lazy_static! {
+    static ref RE_ACCEPT_ENCODING: Regex =
+        Regex::new(r"(?P<a>[\w\*]+)(;q=(?P<q>\d+(\.\d{1,2})?))?(\s*,\s*)?").unwrap();
+}
+
+/// Encodings for Compression/Decompression middleware.
+#[derive(Debug, Clone, Copy, Eq, PartialEq)]
+pub enum Encoding {
+    Gzip,
+    Deflate,
+    Identity,
+    Br,
+}
+
+impl Encoding {
+    fn header_value(self) -> HeaderValue {
+        match self {
+            Encoding::Gzip => HeaderValue::from_str("gzip").unwrap(),
+            Encoding::Deflate => HeaderValue::from_str("deflate").unwrap(),
+            Encoding::Identity => HeaderValue::from_str("identity").unwrap(),
+            Encoding::Br => HeaderValue::from_str("br").unwrap(),
+        }
+    }
+}
+#[derive(Debug, Default)]
+pub struct EncodeSettings {
+    gzip_level: flate2::Compression,
+    deflate_level: flate2::Compression,
+    brotli_level: u32,
```

Comment:
I've changed this in my latest push, since I imagine we have to work with the raw compressor now -- I'm just thinking about how one would be able to use a compressor instance that stays alive with the rest of the stream without the stream taking ownership of it...